cms/server/contest/handlers/contest.py

```
@@ -81,6 +81,8 @@ def __init__(self, *args, **kwargs):
         self.contest: Contest
         self.training_program: TrainingProgram | None = None
         self.impersonated_by_admin = False
+        # Cached eligibility check result to avoid duplicate queries
+        self._eligibility_cache: tuple[bool, "TrainingDayGroup | None", list[str]] | None = None
```

Ruff (0.14.10)

Undefined name `TrainingDayGroup`

(F821)